<a href="https://colab.research.google.com/github/mauriavi/mauriavi.github.io/blob/main/Copy_of_Project5_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NOTE: Please make a copy of this notebook before you start changing anything.

## Loading Data



### Function to fetch data


In [ ]:
def fetch_text(raw_url):

  """
  Fetches text content from a given URL, utilizing a local cache to prevent redundant downloads.

  The function first checks if the content from the `raw_url` is already present in a local cache
  directory (`cs_110_content/text_cache`). If cached, it returns the stored content.
  Otherwise, it sends an HTTP GET request to the URL, saves the fetched content to the cache,
  and then returns it. Error handling is included to catch potential issues during the fetch process.

  Parameters:
    raw_url (str): The URL of the text file to fetch.

  Returns:
    str: The fetched text content if successful.
    None: If an error occurs during fetching or the content cannot be retrieved.

  Notes:
  - The cache directory `cs_110_content/text_cache` is created if it does not already exist.
  - A SHA-1 hash of the URL is used to create a unique filename for caching.
  - A timeout of 10 seconds is applied to the HTTP GET request.
  - `requests.raise_for_status()` is used to raise an HTTPError for bad responses (4xx or 5xx).
  - The function prints a success or failure message to the console.
  """

  import requests
  from pathlib import Path
  import hashlib

  CACHE_DIR = Path("cs_110_content/text_cache")
  CACHE_DIR.mkdir(parents=True, exist_ok=True)

  def _url_to_filename(url):
    url_hash = hashlib.sha1(url.encode("utf-8")).hexdigest()[:12]
    return CACHE_DIR / f"{url_hash}.txt"

  cache_path = _url_to_filename(raw_url)

  SUCCESS_MSG = "✅ Text fetched."
  FAILURE_MSG = "❌ Failed to fetch text."
  try:
    if not cache_path.exists():
      response = requests.get(raw_url, timeout=10)
      response.raise_for_status()
      text_data = response.text
      cache_path.write_text(text_data, encoding="utf-8")
    print(SUCCESS_MSG)
    return cache_path.read_text(encoding="utf-8")

  except Exception as e:
    print(FAILURE_MSG)
    print(f"Error: {e}")
    return ""

### Save the text in a variable

In [ ]:
wuthering_heights_URL = "https://www.gutenberg.org/cache/epub/768/pg768.txt"

wuthering_heights_text = fetch_text(wuthering_heights_URL)

✅ Text fetched.


## Statistics about the data

In [ ]:
import operator

def print_top_10_frequent_words(text):
    """
    Prints the top 10 most frequent words from a given text.

    This function first tokenizes and counts the frequency of each word in the input text
    (by calling `get_word_counts`). It then sorts these word counts in descending order
    and prints the top 10 most frequent words along with their respective counts.

    Parameters:
      text (str): The input text string to analyze.

    Returns:
      None: This function does not return any value; it prints the output directly to the console.

    Notes:
    - Word counting is case-insensitive (all words are converted to lowercase).
    - The definition of a "word" is determined by the `get_word_counts` function,
      which uses whitespace as a delimiter and includes punctuation attached to words.
    - If the text contains fewer than 10 unique words, all unique words will be printed.
    """
    word_counts = get_word_counts(text)
    sorted_word_counts = dict(sorted(word_counts.items(), key=operator.itemgetter(1), reverse=True))
    top_10_words = list(sorted_word_counts.items())[:10]  # Get the top 10 words and counts
    for word, count in top_10_words:
        print(f"{word}: {count}")

print_top_10_frequent_words(wuthering_heights_text)


the: 4692
and: 4409
to: 3532
i: 3216
a: 2343
of: 2329
he: 1832
in: 1436
his: 1393
you: 1381


In [ ]:
#!pip install spacy

In [ ]:
import spacy

In [ ]:
nlp = spacy.load('en_core_web_sm')

In [ ]:
def word_tokenization_normalization(text):
    """
    Performs tokenization and normalization on the input text using spaCy.

    This function processes raw text by converting it to lowercase, tokenizing it,
    and then normalizing the tokens. Normalization includes removing stopwords,
    punctuation, numbers, and words shorter than three characters. The remaining
    words are lemmatized to their base form.

    Parameters:
      text (str): The input text string to be processed.

    Returns:
      list: A list of normalized and lemmatized words (strings) from the input text.

    Notes:
    - The text is first converted to lowercase.
    - Stopwords (common words like 'the', 'is'), punctuation, and numbers are removed.
    - Only words with a length greater than 2 characters after stripping whitespace are kept.
    - Words are returned in their base (lemmatized) form.
    """

    text = text.lower() # lowercase
    doc = nlp(text)     # loading text into model

    words_normalized = []
    for word in doc:
        if word.text != '\n' \
        and not word.is_stop \
        and not word.is_punct \
        and not word.like_num \
        and len(word.text.strip()) > 2:
            word_lemmatized = str(word.lemma_)
            words_normalized.append(word_lemmatized)

    return words_normalized

In [ ]:
doc_tokenized = word_tokenization_normalization(wuthering_heights_text)

In [ ]:
print(doc_tokenized)

['\ufeffthe', 'project', 'gutenberg', 'ebook', 'wuthering', 'height', 'ebook', 'use', 'united', 'states', 'part', 'world', 'cost', 'restriction', 'whatsoever', 'copy', 'away', 'use', 'term', 'project', 'gutenberg', 'license', 'include', 'ebook', 'online', 'www.gutenberg.org', 'locate', 'united', 'states', 'check', 'law', 'country', 'locate', 'ebook', 'title', 'wuthere', 'height', 'author', 'emily', 'brontë', 'release', 'date', 'december', 'ebook', 'recently', 'update', 'language', 'english', 'information', 'format', 'www.gutenberg.org/ebooks/768', 'credit', 'david', 'price', 'start', 'project', 'gutenberg', 'ebook', 'wuthering', 'height', 'wuthering', 'height', 'emily', 'brontë', 'chapter', 'return', 'visit', 'landlord', 'solitary', 'neighbour', 'shall', 'trouble', 'certainly', 'beautiful', 'country', 'england', 'believe', 'fix', 'situation', 'completely', 'remove', 'stir', 'society', 'perfect', 'misanthropist', 'heaven', 'heathcliff', 'suitable', 'pair', 'divide', 'desolation', 'capit

In [ ]:
def word_count(word_list):
    """
    Calculates the frequency of each word in a given list of words.

    This function iterates through a list of words, converts each word to lowercase,
    and counts its occurrences. It returns a dictionary where keys are the unique
    words (in lowercase) and values are their respective counts.

    Parameters:
      word_list (list): A list of strings, where each string is a word.

    Returns:
      dict: A dictionary where keys are words (str) and values are their counts (int).

    Notes:
    - All words are converted to lowercase before counting.
    - The function does not perform any tokenization or cleaning; it assumes the input
      `word_list` already contains individual words.
    """
    word_counts = {}
    for word in word_list:
      word = word.lower()
      if word in word_counts:
        word_counts[word] += 1
      else:
        word_counts[word] = 1
    return word_counts

In [ ]:
new_counts = word_count(doc_tokenized)
print(new_counts)

{'\ufeffthe': 1, 'project': 93, 'gutenberg': 88, 'ebook': 19, 'wuthering': 29, 'height': 106, 'use': 35, 'united': 15, 'states': 12, 'part': 12, 'world': 46, 'cost': 5, 'restriction': 2, 'whatsoever': 2, 'copy': 24, 'away': 90, 'term': 30, 'license': 18, 'include': 16, 'online': 4, 'www.gutenberg.org': 5, 'locate': 7, 'check': 17, 'law': 44, 'country': 25, 'title': 3, 'wuthere': 37, 'author': 2, 'emily': 2, 'brontë': 2, 'release': 10, 'date': 7, 'december': 1, 'recently': 2, 'update': 2, 'language': 9, 'english': 3, 'information': 16, 'format': 6, 'www.gutenberg.org/ebooks/768': 1, 'credit': 5, 'david': 1, 'price': 2, 'start': 24, 'chapter': 38, 'return': 117, 'visit': 50, 'landlord': 8, 'solitary': 6, 'neighbour': 14, 'shall': 200, 'trouble': 38, 'certainly': 16, 'beautiful': 13, 'england': 5, 'believe': 46, 'fix': 24, 'situation': 7, 'completely': 14, 'remove': 38, 'stir': 28, 'society': 15, 'perfect': 9, 'misanthropist': 2, 'heaven': 27, 'heathcliff': 476, 'suitable': 2, 'pair': 6, 

In [ ]:
import operator

def print_top_15_frequent_words(word_counts):
    """
    Prints the top 15 most frequent words from a given word count dictionary.

    This function sorts the word count dictionary in descending order based on
    the word frequencies and then prints the top 15 words along with their counts.

    Parameters:
      word_counts (dict): A dictionary where keys are words (str) and values are their counts (int).

    Returns:
      None: This function does not return any value; it prints the output directly.

    Notes:
    - The function uses the `operator.itemgetter(1)` to sort the dictionary items by their values (counts).
    - It takes a slice of the sorted list to get only the top 15 words.
    - If the dictionary contains fewer than 15 unique words, all words will be printed.
    """
    sorted_word_counts = dict(sorted(word_counts.items(), key=operator.itemgetter(1), reverse=True))
    top_15_words = list(sorted_word_counts.items())[:15]  # Get the top 15 words and counts
    for word, count in top_15_words:
        print(f"{word}: {count}")


In [ ]:
print_top_15_frequent_words(word_counts)

the: 4692
and: 4409
to: 3532
i: 3216
a: 2343
of: 2329
he: 1832
in: 1436
his: 1393
you: 1381
her: 1298
she: 1213
that: 1093
was: 1078
my: 1068


In [ ]:
import operator

def get_verb_counts(text):
    """
    Extracts, tokenizes, normalizes, and counts verbs from the input text using spaCy.

    This function processes raw text by converting it to lowercase, tokenizing it,
    and then normalizing the tokens. Normalization includes identifying verbs,
    removing stopwords, punctuation, numbers, and words shorter than three characters.
    The remaining verbs are lemmatized to their base form and their frequencies are counted.

    Parameters:
      text (str): The input text string to be processed.

    Returns:
      dict: A dictionary where keys are lemmatized verbs (str) and values are their counts (int).
    """
    # Assuming 'nlp' is already loaded globally by spacy.load('en_core_web_sm')
    # If not, it should be loaded here: nlp = spacy.load('en_core_web_sm')

    text = text.lower() # lowercase
    doc = nlp(text)     # loading text into model

    verb_counts = {}
    for token in doc:
        # Check if the token is a verb and meets other normalization criteria
        if token.pos_ == "VERB" \
        and not token.is_stop \
        and not token.is_punct \
        and not token.like_num \
        and len(token.text.strip()) > 2: # Ensure it's a verb and not a short, meaningless token
            lemma = token.lemma_
            verb_counts[lemma] = verb_counts.get(lemma, 0) + 1
    return verb_counts

# Get verb counts
verb_counts = get_verb_counts(wuthering_heights_text)

# Sort the verb counts and get the top 10
sorted_verb_counts = dict(sorted(verb_counts.items(), key=operator.itemgetter(1), reverse=True))
top_10_verbs = list(sorted_verb_counts.items())[:10]

print("Top 10 Most Frequent Verbs:")
for verb, count in top_10_verbs:
    print(f"{verb}: {count}")

Top 10 Most Frequent Verbs:
say: 408
come: 360
think: 236
go: 234
tell: 213
look: 191
know: 186
answer: 181
hear: 163
cry: 157


1. Top 15 Words Findings

the: 4692
and: 4409
to: 3532
i: 3216
a: 2343
of: 2329
he: 1832
in: 1436
his: 1393
you: 1381
her: 1298
she: 1213
that: 1093
was: 1078
my: 1068t

These words do reflect the text's main theme of the love story between a man
and woman. Nothing in the list is suprising or out of the texts context.
Based on the text I can infer that this is a story between a man and a woman based of the pronouns "her, she, he, his". There is also a first person perspective because the use of I in the text is on the top 4 most used. I say the word frequencies capture the text's essence pretty well with the use of his and my, these words feed into the yearning aspect of the text. I was surprised I didnt see any specific names but since the book is mostly about the two main characters the words like she and he did not surpise me. The word was also makes sense for the yearning aspect of something that was theirs but went away.


2. Top 10 Most Frequent Verbs:
say: 408
come: 360
think: 236
go: 234
tell: 213
look: 191
know: 186
answer: 181
hear: 163
cry: 157

Extended Analysis
I did this to gain more context of the text with only the verbs. The verbs give insight into what the text is by listing the action words in the book, for example "cry". These words capture the essence of the book accurately in the context of yearning and the inability to be someone you desperately need. Words like "cry, hear, say, come" all tell a story with the individual verb when placed in the context of yearning.